### DISCUSSION TITLE
Audio Sync in Text to Vid

### DISCUSSION THREAD

**Discussion by syrae** [volotat/SD-CN-Animation#84] [General]
anyone know how to sync vid with the audio on sd cn text to vid?

**Comment by GeekyGhost** [comment]
Are you trying to use extracted onsets or are you just trying to match the video length to the audio length? Beat sync and onset sync are complicated, the latter isn't. I have a concatenate with rhythm feature in my concatenator. I think others have better ones. Deforum has a really good version with Cadence I think.

**Reply by syrae** [reply]
I am after Beat Sync. Soo I've got a 4 min song of which I have keyframes made in https://www.framesync.xyz/. 
I want the animation transition made in SD CN to match the beat of the song. 
Sorry this is a super noob query as I don't have much knowledge about this tech and I am learning. 
For Instance, when I go to SD CN text to vid. Do I just paste the keyframes on the prompts?

**Reply by syrae** [reply]
![wtd](https://github.com/volotat/SD-CN-Animation/assets/67811479/a0d8c528-deeb-4d19-8d56-0f7536cb8e08)

**Comment by volotat** [comment]
Many many years ago I was developing a system that could transform sound into colors. You can see demos here:
https://www.youtube.com/watch?v=QyAEsaX9xzM
https://www.youtube.com/watch?v=rCBFJSVdGFg
https://www.youtube.com/watch?v=JDX8QFWfGuI
It was supposed to be a rithm game that never came. I could use some of those findings and modern ML techniques for audio synchronization, but it is a thing for a future, if the project will stay relevant by then.

**Reply by GeekyGhost** [reply]
That's actually really cool. I've been working on a non ML route, but it requires the right number of frames to go with the number of beats or it doesn't work. I wonder if I could incorporate FILM to pad the frames when needed?

**Reply by GeekyGhost** [reply]
Here's the latest iteration of it lol.

import librosa
import numpy as np

# Load the audio file
y, sr = librosa.load(audio_file)

# Use librosa to find the beats
tempo, beat_frames = librosa.beat.beat_track(y, sr)

# Convert beat frames to time
beat_times = librosa.frames_to_time(beat_frames, sr)

# Calculate the time intervals between beats
beat_intervals = np.diff(beat_times)

# Pad the beat_intervals array to make it the same length as the images array
# This is necessary because there is one less beat interval than there are beats
beat_intervals = np.pad(beat_intervals, (0, len(images) - len(beat_intervals)), 'edge')

# Now, use beat_intervals as the duration for each image
for image, duration in zip(images, beat_intervals):
    img_path = os.path.join(image_folder, image)
    img = mpy.ImageClip(img_path, duration=duration)
    clips.append(img)

# Concatenate video clips
video = mpy.concatenate_videoclips(clips)

# Set the audio
video = video.set_audio(audio_clip)

# Write the result to a file
video.write_videofile(output_file, fps=24)

**Reply by volotat** [reply]
I think yes. But I'm thinking more about the proposal from that topic https://github.com/volotat/SD-CN-Animation/issues/94, particularly we could generated some colorful prime video that is synced to the music via traditional techniques from generative art and then use it as a primer for CN. As you can already use a video as a guide for CN in text to video mode. I think this is the most promising approach with the maximum amount of artistic freedom.

**Reply by GeekyGhost** [reply]
> I think yes. But I'm thinking more about the proposal from that topic #94, particularly we could generated some colorful prime video that is synced to the music via traditional techniques from generative art and then use it as a primer for CN. As you can already use a video as a guide for CN in text to video mode. I think this is the most promising approach with the maximum amount of artistic freedom.

Fair points!
